Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,222 @@
+    }
+  }
+  
+  init(configuration: Configuration) {
+    self.state = .clientIdleServerIdle(configuration)
+  }
+  
+  mutating func send(metadata: Metadata) throws {
+    // Client only sends metadata when opening.
+    try self.clientSentMetadata()
+  }
+  
+  mutating func send(message: [UInt8]) {
+    // Client sends message.
+    switch self.state {
+    case .clientIdleServerIdle(let configuration):
+      preconditionFailure("Client not yet open")
+    case .clientOpenServerIdle(var clientOpenState):
+      clientOpenState.framer.append(message)
+      self.state = .clientOpenServerClosed(clientOpenState)
+    case .clientOpenServerOpen(var clientOpenState):
+      clientOpenState.framer.append(message)
+      self.state = .clientOpenServerClosed(clientOpenState)
+    case .clientOpenServerClosed:
+      // Nothing to do: no point in sending a message if the server's closed.
+      ()
+    case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+      preconditionFailure("Client is closed, cannot send a message")
+    }
+  }
+  
+  mutating func send(status: String, trailingMetadata: Metadata) throws {
+    // Only server does this.
+  }
+  
+  /// Returns the client's next request to the server.
+  /// - Returns: The request to be made to the server.
+  mutating func nextRequest() throws -> ByteBuffer? {
+    switch self.state {
+    case .clientIdleServerIdle(let configuration):
+      preconditionFailure("Client is not open yet.")
+    case .clientOpenServerIdle(var clientOpenState):
+      let request = try clientOpenState.framer.next(compressor: clientOpenState.compressor)
+      self.state = .clientOpenServerIdle(clientOpenState)
+      return request
+    case .clientOpenServerOpen(var clientOpenState):
+      let request = try clientOpenState.framer.next(compressor: clientOpenState.compressor)
+      self.state = .clientOpenServerOpen(clientOpenState)
+      return request
+    case .clientOpenServerClosed:
+      // Nothing to do: no point in sending request if server is closed.
+      return nil
+    case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+      preconditionFailure("Can't send request if client is closed")
+    }
+  }
+  
+  mutating func receive(metadata: Metadata, endStream: Bool) {
+    // This is metadata received by the client from the server.
+    // It can be initial, which confirms that the server is now open;
+    // or an END_STREAM trailer, meaning the response is over.
+    if endStream {
+      self.clientReceivedEndHeader()
+    } else {
+      self.clientReceivedMetadata()
+    }
+  }
+  
+  mutating func receive(message: ByteBuffer, messageProcessor: ([UInt8]) throws -> Void) throws {
+    // This is a message received by the client, from the server.
+    switch self.state {
+    case .clientIdleServerIdle(let configuration):
+      preconditionFailure("Cannot have received anything from server if client is not yet open.")
```

Comment:
We've changed our stance a bit on using `preconditionFailure`, anything that's network reachable should be an assertion failure (i.e. we can discover it in debug mode) but an error in release mode.